src/nvim/change.c

```diff
@@ -523,6 +508,14 @@ void deleted_lines_mark(linenr_T lnum, int count)
 /// @param xtra number of extra lines (negative when deleting)
 void changed_lines_redraw_buf(buf_T *buf, linenr_T lnum, linenr_T lnume, linenr_T xtra)
 {
+  // If lines haven been deleted and there are virtual lines in the buffer,
+  // ensure win_update() calculates the new height of the line to which
+  // virtual lines may have moved to, which could be the line below the
+  // last changed line, so buf->b_mod_bot must be set two lines below.
+  if (xtra < 0 && (buf_meta_total(buf, kMTMetaLines))) {
+    lnume += 2;
+  }
```

Comment:
When `xtra > 1 && buf_meta_total(buf, kMTMetaLines)`, also need to increase `lnume` by `1`.

Reply:
With negative `xtra` extmarks in the deleted lines may move to a different line. With positive `xtra` this doesn't happen.